src/borg/crypto/file_integrity.py

```diff
@@ -0,0 +1,183 @@
+
+    Similarly skipping while writing to create sparse files is also not supported.
+
+    Data has to be read/written in a symmetric fashion, otherwise different
+    digests will be generated.
+
+    Note: When used as a context manager read/write operations outside the enclosed scope
+    are illegal.
+    """
+
+    ALGORITHM = 'SHA-512'
+
+    def __init__(self, backing_fd, write):
+        self.fd = backing_fd
+        self.writing = write
+        self.hash = hashlib.sha512()
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        if exc_type is None:
+            self.hash_length()
+        super().__exit__(exc_type, exc_val, exc_tb)
+
+    def write(self, data):
+        """
+        Write *data* to backing file and update internal state.
+        """
+        n = self.fd.write(data)
+        self.hash.update(data)
+        return n
+
+    def read(self, n=None):
+        """
+        Read *data* from backing file (*n* has the usual meaning) and update internal state.
+        """
+        data = self.fd.read(n)
+        self.hash.update(data)
+        return data
+
+    def digest(self):
+        """
+        Return current digest bytes.
+
+        Note: this can be called multiple times.
+        """
+        return self.hash.digest()
+
+    def update(self, data: bytes):
+        self.hash.update(data)
+
+    def hash_length(self, seek_to_end=False):
+        if seek_to_end:
+            # Add length of file to the hash to avoid problems if only a prefix is read.
+            self.seek(0, io.SEEK_END)
+        self.hash.update(str(self.tell()).encode())
+
+
+class FileIntegrityError(IntegrityError):
+    """File failed integrity check: {}"""
+
+
+class IntegrityCheckedFile(FileLikeWrapper):
+    def __init__(self, path, write, filename=None, override_fd=None):
+        self.path = path
+        self.writing = write
+        mode = 'wb' if write else 'rb'
+        self.file_fd = override_fd or open(path, mode)
+
+        self.fd = self.hasher = SHA512FileHashingWrapper(backing_fd=self.file_fd, write=write)
+
+        self.hash_filename(filename)
+
+        if write:
+            self.digests = {}
+        else:
+            self.digests = self.read_integrity_file(path, self.hasher)
+            # TODO: When we're reading but don't have any digests, i.e. no integrity file existed,
+            # TODO: then we could just short-circuit.
+
+    def hash_filename(self, filename=None):
+        # Hash the name of the file, but only the basename, ie. not the path.
+        # In Borg the name itself encodes the context (eg. index.N, cache, files),
+        # while the path doesn't matter, and moving e.g. a repository or cache directory is supported.
+        # Changing the name however imbues a change of context that is not permissible.
+        filename = os.path.basename(filename or self.path)
+        self.hasher.update(('%10d' % len(filename)).encode())
```

Comment:
oh, i totally misread that (thought it is file length). yeah, it's enough for a filename. :-D